### PR TITLE
support LIMIT OFFSET

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

support OFFSET clause after LIMIT.

## Changelog

- New Feature

## Test Plan

DRAFT and add tests later.

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by TennyZhuang** [comment]
@BohuTANG @sundy-li @zhang2014

File: common/streams/src/stream_limit.rs [hunk 1]
```diff
@@ -14,38 +14,64 @@ use crate::SendableDataBlockStream;
 
 pub struct LimitStream {
     input: SendableDataBlockStream,
-    limit: usize,
+    limit: Option<usize>,
```
Comment (zhang2014) [inline common/streams/src/stream_limit.rs:17]: Maybe it should be `offset: Option<usize>`?

SQL LIMIT syntax:
SELECT * FROM `table` [LIMIT [offset_index, ] limit_size]

Comment (TennyZhuang) [reply common/streams/src/stream_limit.rs:17]: It seems that there is no any differences between offset 0 and no offset clause, so I omit it.

File: common/streams/src/stream_limit.rs [hunk 2]
```diff
@@ -14,38 +14,64 @@ use crate::SendableDataBlockStream;
...
+        let limit = self.limit.unwrap_or(usize::MAX - self.offset);
+        let offset = self.offset;
+        let current = self.current;
+
+        // There are two intervals:
+        // r1: [current, current + rows)
+        // r2: [offset, offset + limit), note that limit may be infinite.
+        //
+        // There are 6 possible relationships between r1 and r2:
+        if current + rows <= offset {
```
Comment (zhang2014) [inline common/streams/src/stream_limit.rs:48]: It's ok, but this code complicated for me. maybe we implement it as `IgnoreHeadStream { ignore_rows: usize }` is better?

for example:
```
SELECT * FROM `table` LIMIT 5, 10; -> LimitStream {limit: 10, input: IgnoreHeadStream {ignore_rows: 5}}
```

Comment (TennyZhuang) [reply common/streams/src/stream_limit.rs:48]: Good idea! And we can rename them to TakeStream and SkipStream.

Comment (TennyZhuang) [reply common/streams/src/stream_limit.rs:48]: I will implement them later.

Comment (TennyZhuang) [reply common/streams/src/stream_limit.rs]: The refactor is done

File: common/streams/src/stream_limit.rs (hunk 1 above)
Comment (TennyZhuang) [reply common/streams/src/stream_limit.rs]: the offset and limit is just like Stream::size_hint(), which use usize representing lower bound and Option<usize> representing upper bound.

File: common/streams/src/stream_skip.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+impl Stream for SkipStream {
+    type Item = Result<DataBlock>;
+
+    fn poll_next(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.input.poll_next_unpin(ctx).map(|x| match x {
+            Some(Ok(ref block)) => Some(Ok({
+                let rows = block.num_rows();
+                if self.current + rows <= self.n {
+                    self.current += rows;
+                    DataBlock::empty_with_schema(block.schema().clone())
```
Comment (zhang2014) [inline common/streams/src/stream_skip.rs:41]: Why do we need to return an empty block? maybe we can eliminate it by looping?

Comment (TennyZhuang) [reply common/streams/src/stream_skip.rs:41]: Fixed, I haven't implement a custom Stream before, but I'm familar with it now!

Comment (zhang2014) [reply common/streams/src/stream_skip.rs:41]: We can update it on the next PR, now let's try to merge this pr.

**Review by zhang2014 — APPROVED** [review]
Thank your contribution.
LGTM.

File: common/streams/src/stream_skip.rs (hunk 3 above)
Comment (TennyZhuang) [reply common/streams/src/stream_skip.rs:41]: I've finished it already but failed to commit due to conflict 😓

Comment (zhang2014) [reply common/streams/src/stream_skip.rs]: Sorry, I tried to modify it to pass the cluster test 😓

**Comment by TennyZhuang** [comment]
> Thank your contribution.
> LGTM.

I've pushed a simple rewrite for StreamSkip using loop to avoid meaningless empty datablock, sorry but please review it again! Thanks.

**Comment by TennyZhuang** [comment]
> > Thank your contribution.
> > LGTM.
> 
> I've pushed a simple rewrite for StreamSkip using loop to avoid meaningless empty datablock, sorry but please review it again! Thanks.

See e239c61925d2ca3d5717c2e2cbfa2564e47b2b4c

File: common/streams/src/stream_skip.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            remaining: n,
+        }
+    }
+}
+
+impl Stream for SkipStream {
+    type Item = Result<DataBlock>;
+
+    fn poll_next(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        while self.remaining > 0 {
```
Comment (zhang2014) [inline common/streams/src/stream_skip.rs:34]: 👍

Comment (BohuTANG) [reply common/streams/src/stream_skip.rs:34]: 👍

**Pull request merged**